Routing modularization proposal

xavierraffin opened the pull request:
We aim to make Jormungandr more flexible.
This is a functional (almost) non intrusive modification which add a module system to Jormungandr.

All modules are now stored in modules directory.

A module have a namespace which is an url prefix.
That avoids route collision between modules

Navitia.io API is now a module prefixed by v1

The Jormungandr behaviour is unchanged

Future improvements
- generate dynamicaly the Jormungandr entry point http://api.navitia.io with modules information
- add some other example modules ?

buildbotctp commented:
Can one of the admins verify this patch?

antoine-de commented:
jenkins, test this please

antoine-de commented:
thanks @xavierraffin we're looking into it

antoine-de commented on source/jormungandr/jormungandr/modules/__init__.py:
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env python
```
you shouldn't put those 2 lines.

the file should not be in utf-8 and is not executable as is.

antoine-de commented on source/jormungandr/jormungandr/modules/v1_routing.py:
```diff
@@ -0,0 +1,222 @@
+from werkzeug.routing import BaseConverter, FloatConverter, PathConverter
+from jormungandr.modules_loader import AModule
+
+__author__ = 'Nicolas Garcia <nicolas.garcia@tisseo.fr>'
```
we usually don't put a **author** tag in the file, I don't think they're relevant.

antoine-de commented on source/jormungandr/jormungandr/modules/v2_routing.py:
```diff
@@ -0,0 +1,42 @@
+__author__ = 'Nicolas Garcia <nicolas.garcia@tisseo.fr>'
+
+
+class V2Routing(AModule):
```
I think v2 should be for the next 'classic' api version.

what about a 'tisseo' end point ?

`like https://api.navitia.io/v1/...`
`like https://api.navitia.io/v2/...`
`like https://api.navitia.io/tisseo/...` 
? (it is open to debate, not sure about it)

nicolas-garcia commented:
Hi,
About the modularity, I've two questions :
First, in end_point_tests, you check if the number of versions in the endpoint is equal to 1.
But this test became deprecated due to the modularization, right ?

Do I have to remove this assertion ?

Finally, still in the same test, there is an assertion about endpoints status which can only be 'current' or 'deprecated', but this assertion isn't relevant in the case where we have the modularization.

What do I have to do with those tests ?

antoine-de commented:
Hi @nicolas-garcia,

you're right, those tests have to be updated following the modifications (and we'll have to find a new status for your api)

nicolas-garcia commented:
Thanks for your quick answer.

Due to the modularity of endpoints, the solution might be to remove the assertion 
`assert len(versions) == 1`, 
or do you have an other solution ?

antoine-de commented:
you can maybe update it with 
`assert len(versions) >= 1`

and check that at least one is the 'current'

Commit:
Jormungandr routing is now modular:
All route files are organised into modules in a "modules" directory.
Modules could be blueprint if need specific templates and static directories.

There is a example in modules which is desactivated by default.

Now the root index is generated dynamicaly instead of hard coded.

Test are updated accordingly

xavierraffin commented:
The pull request has been updated

TeXitoi commented:
jenkins, test this please

kinnou02 commented:
jenkins, test this please

xavierraffin commented:
So, do you will merge this pull-request ?

kinnou02 commented on source/jormungandr/jormungandr/instance.py:
```diff
@@ -81,12 +82,16 @@ def scenario(self):
             logger = logging.getLogger(__name__)
             instance_db = self._get_models()
             if instance_db:
-                logger.info('loading of scenario %s for instance %s', instance_db.scenario, self.name)
```
we use a line length at  something like 110 characters, it must readable in github without scrolling

nicolas-garcia replied:
I note , I just applied pep8 standard strictly .
I'll fix that. Anyway, that file should not have been modified by the commit.

kinnou02 commented on source/jormungandr/jormungandr/api.py:
```diff
@@ -46,5 +48,16 @@ def output_jsonp(data, code, headers=None):
+                                      'Current version of navitia API',
+                                      status='current'))
+# Uncomment the following lines if you want to activate the test module
+#rest_api.module_loader.load(TisseoRouting(rest_api, 'tisseo/test',
```
Do we really need to set the description and the status at the instantiation? I think it can be statically set in the constructor since it's a property of the module, we don't want to change it without a code modification

kinnou02 replied:
The list of enabled modules must be configurable with the settings.py file, something like this:

```
from jormungandr.modules.tisseo_routing.TisseoRouting import TisseoRouting
from jormungandr.modules.v1_routing.v1_routing import V1Routing
modules = {'v1': V1Routing, 'tisseo/test':  TisseoRouting}
```

then

```
for prefix, module in modules.iteritems():
    rest_api.module_loader.load(module(rest_api, prefix))
```

Commit:
Enabled modules are now configurable in settings.py. Revert formating on instance.py. Module status and description are now in module constructor

nicolas-garcia commented:
The PR has been updated according to previous comments of @kinnou02

antoine-de commented on source/jormungandr/jormungandr/api.py:
```diff
@@ -46,5 +46,15 @@ def output_jsonp(data, code, headers=None):
+        module = getattr(module_file, module_info['class_name'])
+        rest_api.module_loader.load(module(rest_api, prefix))
+
+    rest_api.module_loader.run()
```
since nothing will be done if empty, it might be good to log in that case

antoine-de commented on source/jormungandr/jormungandr/module_resource.py:
```diff
@@ -0,0 +1,65 @@
+        """
+        if not isinstance(resource, ModuleResource):
+            raise TypeError('Expected type {0}, got {1} instead'
+                            .format('ModuleResource',
```
if ModuleResource is hard coded, you can put it directly in the string no ?

Commit:
Minor fixes. Added log

kinnou02 commented:
jenkins, test this please

buildbotctp commented:
thanks @xavierraffin  and @nicolas-garcia !
